File: src/api/discussion-send-message.ts
import { type Discussion, DiscussionActionError, DiscussionMessagesMissingError, EntityState, type SessionHandle, TabLocation } from "~/models";
import { encodeDiscussionSendAction } from "~/encoders/discussion-send-action";
import { RequestFN } from "~/core/request-function";
import { discussionMessages } from "./discussion-messages";
import { createEntityID } from "./helpers/entity-id";
import { discussions } from "./discussions";

export const discussionSendMessage = async (
  session: SessionHandle,
  discussion: Discussion,
  content: string,
  includeParentsAndStudents = false,
  replyTo = discussion.messages?.defaultReplyMessageID
): Promise<void> => {
  if (!discussion.messages)
    throw new DiscussionMessagesMissingError();

  if (typeof discussion.messages.sendAction === "undefined")
    throw new DiscussionActionError();

  const action = encodeDiscussionSendAction(discussion.messages.sendAction, includeParentsAndStudents);
  const request = new RequestFN(session, "SaisieMessage", {
    donnees: {
      contenu: session.user.authorizations.hasAdvancedDiscussionEditor ? {
        _T: 21,
        V: content
      } : content,

      bouton: { N: 0, G: action },
      brouillon: {
        N: createEntityID(),
        E: EntityState.CREATION
      },
      genreDiscussion: 0,
      messagePourReponse: {
        G: 0,
        N: replyTo
      },

      listeFichiers: []
    },

    _Signature_: {
      onglet: TabLocation.Discussions
    }
  });

  await request.send();

  await discussions(session, discussion.cache);
  await discussionMessages(session, discussion);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/discussion-send-message.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.helpers.createEntityID
import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.encoders.encodeDiscussionSendAction
import ink.literate.pawnote.models.Discussion
import ink.literate.pawnote.models.EntityState
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import ink.literate.pawnote.models.errors.DiscussionActionError
import ink.literate.pawnote.models.errors.DiscussionMessagesMissingError
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun discussionSendMessage (
    session: SessionHandle,
    discussion: Discussion,
    content: String,
    includeParentsAndStudents: Boolean = false,
    replyTo: String? = discussion.messages?.defaultReplyMessageID
) {
    if (discussion.messages == null)
        throw DiscussionMessagesMissingError()

    if (discussion.messages!!.sendAction == null)
        throw DiscussionActionError()

    val action = encodeDiscussionSendAction(discussion.messages!!.sendAction!!, includeParentsAndStudents)

    val request = RequestFN(session.information, "SaisieMessage", Json.encodeToString(
        buildJsonObject {
            putJsonObject("donnees") {
                if (session.user.authorizations.hasAdvancedDiscussionEditor)
                    putJsonObject("contenu") {
                        put("_T", 21)
                        put("V", content)
                    }
                else
                    put("contenu", content)

                putJsonObject("bouton") {
                    put("N", 0)
                    put("G", action.code)
                }

                putJsonObject("brouillon") {
                    put("N", createEntityID())
                    put("E", EntityState.CREATION.code)
                }

                put("genreDiscussion", 0)

                putJsonObject("messagePourReponse") {
                    put("G", 0)
                    put("N", replyTo)
                }

                putJsonArray("listeFichiers") {}
            }

            putJsonObject("_Signature_") {
                put("onglet", TabLocation.Discussions.code)
            }
        }
    ))

    request.send()

    discussions(session, discussion.cache)
    discussionMessages(session, discussion)
}